src/formats/CIF.cpp

```diff
@@ -0,0 +1,427 @@
+            Residue residue(name.to_string(), resid);
+            residue.add_atom(atom_id);
+
+            // This will be saved as a string on purpose to match MMTF
+            if (label_asym_id != atom_site_map_.end()) {
+                residue.set("chainid", chainid);
+            }
+
+            if (auth_asym_id != atom_site_map_.end()) {
+                residue.set("chainname", line_split[auth_asym_id->second].to_string());
+            }
+
+            if (group_pdb != atom_site_map_.end()) {
+                residue.set("is_standard_pdb", line_split[group_pdb->second] == "ATOM");
+            }
+
+            residues_.emplace(std::make_pair(chainid, resid), std::move(residue));
+        } else {
+            // Just add this atom to the residue
+            residues_.at({chainid, resid}).add_atom(atom_id);
+        }
+    }
+
+    // Reset state to previous line
+    file_.seekpos(position);
+
+    for (const auto& residue: residues_) {
+        frame.add_residue(residue.second);
+    }
+
+    // Only link if we are reading CIF
+    if (model_position != atom_site_map_.end()) {
+        // Cross format talk! Forgive me!
+        PDBFormat::link_standard_residue_bonds(frame);
+    }
+}
+
+void CIFFormat::write(const Frame& frame) {
+    file_.print("data_model_{}\n", models_); // TODO: do models have a name in Chemfiles? If so, use it
+    file_.print("_audit_creation_method         'generated by Chemfiles'\n");
+    file_.print("_symmetry_cell_setting         'triclinic'\n");
+    file_.print("_symmetry_space_group_name_H-M 'P 1'\n");
+    file_.print("_cell.length_a {}\n", frame.cell().a());
+    file_.print("_cell.length_b {}\n", frame.cell().b());
+    file_.print("_cell.length_c {}\n", frame.cell().c());
+    file_.print("_cell.length_alpha {}\n", frame.cell().alpha());
+    file_.print("_cell.length_beta  {}\n", frame.cell().beta());
+    file_.print("_cell.length_gamma {}\n", frame.cell().gamma());
+    file_.print("loop_\n");
+    file_.print("  _symmetry_equiv_pos_as_xyz\n");
+    file_.print("  '+x,+y,+z'\n");
+
+    file_.print("\n");
+    file_.print("loop_\n");
+    file_.print("_atom_site_label\n");
+    file_.print("_atom_site_type_symbol\n");
+    file_.print("_atom_site_occupancy\n");
+    file_.print("_atom_site_fract_x\n");
+    file_.print("_atom_site_fract_y\n");
+    file_.print("_atom_site_fract_z\n");
+    file_.print("_atom_site_Cartn_x\n"); // TODO: having Cartn is rare for pure CIF formats, but is allowed. I do not think it hurts.
+    file_.print("_atom_site_Cartn_y\n");
+    file_.print("_atom_site_Cartn_z\n");
+
+
+    const auto& positions = frame.positions();
+    for (size_t i = 0; i < frame.size(); ++i) {
+        const auto& atom = frame[i];
+
+	// TODO: are the fractional coordinates stored? if not, use the cell transformation matrix to transform positions[i] into a fract vector
+	// fract = 
+
+	// TODO: can atoms have nonunit occupancy in Chemfiles? if so, use it here instead of "1.0"
```

Comment:
not yet, but such occupancy could be stored as a new Atom property, read in by all format supporting it and checked when writing.

That's how we deal with PDB insertion code or ATOM/HETATM distinction for examle